[tensorexpr] Linear scan intermediate buffers to reuse memory

huiguoo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66967
* __->__ #66452
* #67144
* #67143
* #66451
* #66450



Differential Revision: [D31557188](https://our.internmc.facebook.com/intern/diff/D31557188)

huiguoo committed:
[tensorexpr] Add memory planning to reuse intermediate buffers

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557188).

huiguoo committed:
Update on "[tensorexpr] Add memory planning to reuse intermediate buffers"




Differential Revision: [D31557188](https://our.internmc.facebook.com/intern/diff/D31557188)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557188).

huiguoo committed:
Update on "[tensorexpr] Add memory planning to reuse intermediate buffers"




Differential Revision: [D31557188](https://our.internmc.facebook.com/intern/diff/D31557188)

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Add memory planning to reuse intermediate buffers"




Differential Revision: [D31557188](https://our.internmc.facebook.com/intern/diff/D31557188)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557188).

huiguoo committed:
Update on "[tensorexpr] Linear scan intermediate buffers to reuse memory"




Differential Revision: [D31557188](https://our.internmc.facebook.com/intern/diff/D31557188)

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Linear scan intermediate buffers to reuse memory"




Differential Revision: [D31557188](https://our.internmc.facebook.com/intern/diff/D31557188)

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Linear scan intermediate buffers to reuse memory"




Differential Revision: [D31557188](https://our.internmc.facebook.com/intern/diff/D31557188)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557188).

huiguoo committed:
Update on "[tensorexpr] Linear scan intermediate buffers to reuse memory"




Differential Revision: [D31557188](https://our.internmc.facebook.com/intern/diff/D31557188)

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Linear scan intermediate buffers to reuse memory"




Differential Revision: [D31557188](https://our.internmc.facebook.com/intern/diff/D31557188)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557188).

huiguoo committed:
Update on "[tensorexpr] Linear scan intermediate buffers to reuse memory"




Differential Revision: [D31557188](https://our.internmc.facebook.com/intern/diff/D31557188)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557188).

huiguoo committed:
Update on "[tensorexpr] Linear scan intermediate buffers to reuse memory"




Differential Revision: [D31557188](https://our.internmc.facebook.com/intern/diff/D31557188)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557188).

huiguoo committed:
Update on "[tensorexpr] Linear scan intermediate buffers to reuse memory"




Differential Revision: [D31557188](https://our.internmc.facebook.com/intern/diff/D31557188)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557188).

ZolotukhinM commented on test/cpp/tensorexpr/test_memplanning.cpp:
```diff
@@ -0,0 +1,220 @@
+}
+
+#ifdef TORCH_ENABLE_LLVM
+TEST(MemPlanning, MemReuse1onLLVM) {
```
Instead of copying the entire tests, can we just call two codegens (maybe we need to clone stmt for the second one, but it still would be much smaller test)?

ZolotukhinM commented on torch/csrc/jit/tensorexpr/codegen.cpp:
```diff
@@ -83,21 +83,134 @@ void CodeGen::call_with_numel(void** args, int64_t numel) {
+    // Check whether there are free memories that this buf can reuse.
+    for (auto it = mm_free.begin(); it != mm_free.end(); it++) {
+      auto m = *it;
+      TORCH_INTERNAL_ASSERT(bufSize(buf) != 0);
```
This already has been checked on line 155, no need to do it again.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/codegen.cpp:
```diff
@@ -83,21 +83,134 @@ void CodeGen::call_with_numel(void** args, int64_t numel) {
+    for (auto it = mm_free.begin(); it != mm_free.end(); it++) {
+      auto m = *it;
+      TORCH_INTERNAL_ASSERT(bufSize(buf) != 0);
+      TORCH_INTERNAL_ASSERT(bufSize(m) != 0);
```
Should we be comparing against `c10::nullopt` instead?

ZolotukhinM commented on torch/csrc/jit/tensorexpr/codegen.cpp:
```diff
@@ -83,21 +83,134 @@ void CodeGen::call_with_numel(void** args, int64_t numel) {
+      auto m = *it;
+      TORCH_INTERNAL_ASSERT(bufSize(buf) != 0);
+      TORCH_INTERNAL_ASSERT(bufSize(m) != 0);
+      if (bufSize(m) >= bufSize(buf)) {
```
Nit: save bufSize(m) and bufSIze(buf) in variables to avoid recomputing them again.

ZolotukhinM replied:
How does this comparison work if `bufSize(m)` returns `c10::nullopt`? Should we exclude such bufs from `mem_up_for_grabs`?

ZolotukhinM commented on torch/csrc/jit/tensorexpr/codegen.cpp:
```diff
@@ -83,21 +83,134 @@ void CodeGen::call_with_numel(void** args, int64_t numel) {
+  };
+
+  // Map intermediate buffers to the most recent used memory if any.
+  std::unordered_set<BufPtr> mm;
+  std::list<BufPtr> mm_free;
+  std::unordered_map<BufPtr, BufPtr> b2m;
+  std::vector<std::pair<BufPtr, BufPtr>> b2m_ret;
```
Nit: Let's use more descriptive names for variables :)

ZolotukhinM commented on torch/csrc/jit/tensorexpr/codegen.cpp:
```diff
@@ -83,21 +83,134 @@ void CodeGen::call_with_numel(void** args, int64_t numel) {
+    // Release memory for buffers whose live range ends before the creation time
+    // of this buf.
+    // TODO: optimize in-place opererations and copy operations
+    buf_to_release.clear();
+    for (auto& mapped : b2m) {
+      auto buf_mapped = mapped.first;
```
Nit: maybe it's because of the var names, but it's hard for me to follow this code. I don't understand what `mapped` holds, what's in `buf_mapped`, etc. Would it be possible to use more descriptive names and add some comments to make the code more readable?

ZolotukhinM commented on torch/csrc/jit/tensorexpr/codegen.cpp:
```diff
@@ -83,21 +83,134 @@ void CodeGen::call_with_numel(void** args, int64_t numel) {
+      b->prepend_stmt(alloc<Allocate>(buf));
+      b->append_stmt(alloc<Free>(buf));
+    } else {
+      b->prepend_stmt(alloc<PlacementAllocate>(rit->first, rit->second));
```
We're inserting PlacementAllocate to the beginning of the block. I think it's technically correct with the current implementation, but it breaks a neat mental model that it is equivalent to `"Free(old); Allocate(new);"`. Would it be possible to insert it after the last use of the old buffer and before the first use of the new one?

huiguoo replied:
As we discussed in person, we'll move all `Allocate/PlacementAllocate/Free` from the beginning to where they should be in a separate PR.

ZolotukhinM reviewed (commented):
I have some comments inline. Overall the code looks good, but I think we can try to make it a bit cleaner.

huiguoo commented on torch/csrc/jit/tensorexpr/codegen.cpp:
```diff
@@ -111,15 +228,27 @@ void CodeGen::allocIntermediateBufs() {
-    auto stmt_new = insertMemNodes(interm_bufs, stmt_);
+  // For each intermediate buffer, we reuse the memory of an old buffer which
+  // dies, or allocate memory if reusing buffer is impossible.
+  auto b2m = linearScan(interm_bufs, interm_buf_ranges);
```
It is the map between buffers to memory blocks. Fixed the name.

ZolotukhinM replied:
It still reads "b2m", have you pushed the recent changes?

huiguoo committed:
Update on "[tensorexpr] Linear scan intermediate buffers to reuse memory"




Differential Revision: [D31557188](https://our.internmc.facebook.com/intern/diff/D31557188)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/tensorexpr/codegen.cpp:
```diff
@@ -83,21 +83,135 @@ void CodeGen::call_with_numel(void** args, int64_t numel) {
+}
+
+// This algorithm takes the list of intermediate buffers and their liveness
+// ragnes, and returns a map of these buffers to memory blocks. A memory block
+// is indicated by the buffer for which it is allocated. Specifically, we
```
What is a memory block? How different is it from Buf? If it's the same, can we avoid introducing a new concept?

ZolotukhinM commented on torch/csrc/jit/tensorexpr/codegen.cpp:
```diff
@@ -83,21 +83,138 @@ void CodeGen::call_with_numel(void** args, int64_t numel) {
+
+std::vector<std::pair<BufPtr, BufPtr>> linearScan(
+    std::unordered_set<BufPtr>& bufs,
+    BufRangeInfo& buf_ranges) {
```
I would still suggest using a name that reflects "what" rather then "how". Imagine someone reading the code later on and seeing "b2m = linearScan(...);". How would they know what's goind on and why we're doing that?

ZolotukhinM commented on torch/csrc/jit/tensorexpr/codegen.cpp:
```diff
@@ -83,21 +83,135 @@ void CodeGen::call_with_numel(void** args, int64_t numel) {
+
+  // Map intermediate buffers to the most recently used memory if any.
+  std::list<BufPtr> mem_up_for_grabs;
+  std::unordered_map<BufPtr, BufPtr> curr_buf_mem_map;
+  std::vector<std::pair<BufPtr, BufPtr>> global_buf_mem_map;
```
What's the difference between "curr" and "global" here? It seems that the name is misleading and the actual meaning is somewhat different - i.e. one of them is simply a vector while the other is a map. Am I missing something?

huiguoo replied:
`curr_buf_mem_map` holds the mapping of alive buffers to memories at the moment; and `global_buf_mem_map` holds all the mappings as a vector so we can insert `Alloc/PlacementAlloc` later with it with determined order. Fixing the names.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/codegen.cpp:
```diff
@@ -111,15 +225,27 @@ void CodeGen::allocIntermediateBufs() {
+
+      // Identify the access stmts to each unallocated intermeiate buffer.
+      auto range = BufLiveRange::liveRange(stmt_, buf);
+      // TORCH_INTERNAL_ASSERT(accesses.size() >= 1);
+      // auto range =
+      //    std::make_pair(accesses.at(0), accesses.at(accesses.size() - 1));
```
Leftover.

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557188).

huiguoo committed:
Update on "[tensorexpr] Linear scan intermediate buffers to reuse memory"




Differential Revision: [D31557188](https://our.internmc.facebook.com/intern/diff/D31557188)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557188).

ZolotukhinM commented on test/cpp/tensorexpr/test_bufliverange.cpp:
```diff
@@ -1,87 +0,0 @@
-#include <gtest/gtest.h>
-
-#include <stdexcept>
-#include "test/cpp/tensorexpr/test_base.h"
-
-#include <torch/csrc/jit/tensorexpr/analysis.h>
-#include <torch/csrc/jit/tensorexpr/expr.h>
-#include <torch/csrc/jit/tensorexpr/ir.h>
```
Why are we removing this test after we added it just a couple commits earlier?

huiguoo replied:
Oh, the test is moved to file `test_memplanning.cpp`. We only use `BufLiveRange` in memory planning so it makes more sense to put it there.

ZolotukhinM replied:
Code moves likes this make it difficult to review since they camouflage changes in the file, please try to avoid this in future. When in the same stack, why not to put the tests into this file from the beginning, and when it's for existing code, then it's better to move code in a separate PR without other changes.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/codegen.cpp:
```diff
@@ -83,21 +83,136 @@ void CodeGen::call_with_numel(void** args, int64_t numel) {
+}
+
+// This algorithm takes the list of intermediate buffers and their liveness
+// ragnes, and returns the allocations of these buffers. A buffer 'A' can be
```
Typo: s/ragnes/ranges/

ZolotukhinM commented on torch/csrc/jit/tensorexpr/codegen.cpp:
```diff
@@ -83,21 +83,136 @@ void CodeGen::call_with_numel(void** args, int64_t numel) {
+        buf_to_release.end(),
+        sorting_function_by_end_time);
+    for (auto& buf_rl : buf_to_release) {
+      mem_up_for_grabs.push_front(buf_mem_map[buf_rl]);
```
Nit: use `buf_mem_map.at(buf_rl)` instead of `buf_mem_map[buf_rl]`. The difference is that the latter one would insert a null element if it was absent in the map whereas the former one would raise an error.

ZolotukhinM commented on test/cpp/tensorexpr/test_memplanning.cpp:
```diff
@@ -0,0 +1,213 @@
+  BufHandle AP("A", {M, K}, kFloat);
+  BufHandle BP("B", {K, N}, kFloat);
+
+  Tensor CT = Reduce(
+      "gemm",
+      {{M, "M"}, {N, "N"}},
+      Sum(),
+      [&](const ExprHandle& m, const ExprHandle& n, const ExprHandle& k) {
+        return AP.load(m, k) * BP.load(k, n);
+      },
+      {{K, "K"}});
+  Tensor DT = Compute(
+      "relu",
+      {{M, "M"}, {N, "N"}},
+      [&](const ExprHandle& m, const ExprHandle& n) {
+        auto zero = Cast::make(CT.buf()->dtype(), 0);
+        return CompareSelect::make(
+            CT.load(m, n), zero, zero, CT.load(m, n), kLT);
+      });
+  Tensor ET = Compute(
+      "add",
+      {{M * 2, "EM"}, {N * 2, "EN"}},
+      [&](const ExprHandle& em, const ExprHandle& en) {
+        return DT.load(em / 2, en / 2) + DT.load(em / 2, en / 2);
+      });
+  Tensor FT = Compute(
+      "mul",
+      {{M * 2, "FM"}, {N * 2, "FN"}},
+      [&](const ExprHandle& fm, const ExprHandle& fn) {
+        return ET.load(fm, fn) + ET.load(fm, fn);
+      });
+  auto stmt = Block::make({CT.stmt(), DT.stmt(), ET.stmt(), FT.stmt()});
```
It's difficult to wrap one's head around what's going on in the test. Do you mind adding a comment to show how the original IR looks like and why there is no memory reuse (and ditto in the previous test)?

ZolotukhinM commented on torch/csrc/jit/tensorexpr/codegen.cpp:
```diff
@@ -111,15 +226,25 @@ void CodeGen::allocIntermediateBufs() {
+  // For each intermediate buffer, we reuse the memory of an old buffer whose
+  // liveness range does not overlap with the current buffer, or allocate memory
+  // if reusing buffer is impossible.
+  auto buf_allocs = linearScan(interm_bufs, interm_buf_ranges);
```
Nit: The name should tell what the function does, not how it does it.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/codegen.cpp:
```diff
@@ -83,21 +83,136 @@ void CodeGen::call_with_numel(void** args, int64_t numel) {
+         it++) {
+      auto m = *it;
+      if (bufSize(m) >= bufSize(buf)) {
+        buf_mem_map[buf] = m;
```
Say we reused buf 'A' for buf 'B'. How do we mark 'A' from being reused again later? Could you please add a test to check that it works correctly?

I.e. if input IR is like this:
```
A = alloc(10);
B = alloc(10);
C = alloc(10);
first_use(A);
last_use(A)
first_use(B)
first_use(C)
last_use(C)
last_use(B)
Free(A)
Free(B)
Free(C)
```
We need to make sure we do not assign both B and C to A:
```
A = alloc(10);
B = alias(A);
C = alloc(10);
first_use(A);
last_use(A)
first_use(B)
first_use(C)
last_use(C)
last_use(B)
Free(A)
Free(C)
```

huiguoo replied:
All buffers that can be reused at the moment are in the list of `mem_up_for_grabs`. Once buf `A` is reused for buf `B`, buf `A` will be removed from `mem_up_for_grabs`. When we arrive the end of buf `B`'s  liveness range, buf `A` will be put back to `mem_up_for_grabs` for future reuse.

I'll add a test for this.

huiguoo replied:
> How does this comparison work if bufSize(m) returns c10::nullopt? Should we exclude such bufs from mem_up_for_grabs?

We will not put buffers with c10::nullopt bufSize to mem_up_for_grabs. Currently, we only reuse buffers with static shapes. For buffers with dynamic shapes, we will mark it as `allocate` in the following code:
```
     // If the buf has dynamic shapes, we'll skip it (i.e., allocate memory for
     // it, and there are no future reuses on its memory).
     // TODO: reuse memory for bufs with dynamic shapes
     if (!bufSize(buf)) {                                                                                                                                                      
       buf_allocs.emplace_back(std::make_pair(buf, buf));
       continue;
     }
```

huiguoo committed:
Update on "[tensorexpr] Linear scan intermediate buffers to reuse memory"




Differential Revision: [D31557188](https://our.internmc.facebook.com/intern/diff/D31557188)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557188).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/codegen.cpp:
```diff
@@ -83,21 +83,136 @@ void CodeGen::call_with_numel(void** args, int64_t numel) {
+    // Release memory for buffers whose liveness range ends before the creation
+    // time of this buf.
+    // TODO: optimize in-place opererations and copy operations
+    buf_to_release.clear();
```
Nit: instead of clearing we could just declare `buf_to_release` here - this would make it more obvious that its value is only needed within a single iteration of this loop.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/codegen.cpp:
```diff
@@ -83,21 +83,136 @@ void CodeGen::call_with_numel(void** args, int64_t numel) {
+    // If the buf has dynamic shapes, we'll skip it (i.e., allocate memory for
+    // it, and there are no future reuses on its memory).
+    // TODO: reuse memory for bufs with dynamic shapes
+    if (!bufSize(buf)) {
```
Nit: this check could be done in the very beginning so that it's not mixing with the rest of the logic.

ZolotukhinM reviewed (approved):
Looks good! Thanks for addressing all the remarks!

huiguoo committed:
Update on "[tensorexpr] Linear scan intermediate buffers to reuse memory"




Differential Revision: [D31557188](https://our.internmc.facebook.com/intern/diff/D31557188)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557188).
